### ISSUE TITLE
Change CMS permission model for users and admins

### ISSUE THREAD

**Issue by ericvaandering** [dmwm/CMSRucio#343]

**Comment by ericvaandering** [comment]
https://github.com/dmwm/CMSRucio/issues/297

**Comment by belforte** [comment]
Given that no other use case than the CRAB one popped up. What can we do to get it done and close ?
Is this something that also happens via the permission plugins which @dynamic-entropy is using for controlling user Rule creation ?
If so, is this something were we CRAB people could contribute in O(1 week) so that it stops being kicked from one quarter to the next ?

**Comment by dynamic-entropy** [comment]
Hi Stefano
If you refer to #297 linked above, it seems too long and hard to follow. Can I rely on the summary by @dciangot in the description to comment?
Irrespective, I see nothing wrong with allowing users to mark bad replicas in their own scope. 
Let me know, and sorry for the delay. 

For the rule creation permission, we had to change the approach since it became impossible to do it without messing up the already existing "some" user quotas ( their existence in the first place seems wrong, other than to developers, it also messes up so many things on our side, but its all upon us, so I cannot complain, but we can discuss this some other time in some other thread).

**Comment by belforte** [comment]
thanks Rahul, sory for long thread. Yes, the [summary from Diego](https://github.com/dmwm/CMSRucio/issues/297#issuecomment-1274849341) is correct. It quickly describes the solution to [the requirement](https://github.com/dmwm/CMSRucio/issues/297#issuecomment-1273384693).
Being able to declare a replica as bad should be enough.

 Anyhow, since you ask good questions, and it is good to question the "why of things" and look for better, cleaner, safer ways.. let me make an attempt to clarify:
 1. A very large fraction of CRAB workflows produce root ntuple files which users will process locally, so they like/need human readable file names. We can not use GUID like P&R does. In CRAB2 we were still using unique names by adding a retry number suffix (myout_1_1.root myout_2_1.root myout_3.2.root etc.) but it had its own problems, often people ended up with having both myout_154_1.root and myout_154_2.root on disk and could not easily tell what to keep and what to remove.
 2. so we have to stick with: filename can not be changed when resubmitting
 3. the initial replica of those output files is on /store/temp and differently from /store/unmerged this is scratch space where files can be removed after some time even w/o telling Rucio. Also we can have P&R workflows which take months to complete, but users are not that patient, nor CRAB can keep things around that long. So we have the need to declare a transfer as failed and give up on it, even if Rucio would keep trying forever and forever failing.
 
If **we**, which means "a process which authenticates on Rucio with the user credential (X509 proxy)", can declare the replica on temp as BAD, it should be sufficient.
Additional capability to delete a replica may help to make things clean for the case when we give up on waiting on a transfer and do not/can not resubmit the job (e.g. the CRAB task has reached its end-of-life). All in all this is what CompOps operators do, right ?

Finally, I do not understand point 2 in your last comment, unless you mean : the only way to change metadata like checksum on an existing LFN is when all its replicas are BAD (and of course have proper permission ).

P.S. technically we can also run those processes with some CRAB service credential instead of the user ones, but currenly one need Rucio admin privilege to declare replica bad, and that privilege is too much powerful. 

Sorry for long comment. Feel free to ask more in MatterMost, or we can have quick chat

**Comment by dynamic-entropy** [comment]
Hello, So we are talking about two things here, in rucio terminology. 
`DID` (which is what I mean when I say a `file`).
`Replica` (Copies of that file in various RSEs)

A DID gets its checksum, size and other metadata at the time of registration into Rucio. And that becomes its `correct` checksum for its entire life. 
A replica is bad if its checksum does not match this `correct` checksum. 

We cannot change the metadata of the DID after this initial registration (Which happens when you do add_replica for the first time). 
This cannot be done even if you remove the DID and try to register it again. (Which again is how it is supposed to be, a `DID` has to be unique at all times).

----
### Extra Info
There is a bug in the `add_replicas` API which will return `True` (wrongly) even if you try this illegal operation, but it will not actually do anything.
There is another (or can be considered as same) bug, that allows you to register it in another RSE (which is what Diego discovered from what I see in that thread #297 . But as I said it is a Bug, not a solution), but then all your transfers will fail since your initial DID checksum will remain what was registered in the beginning.

**Comment by belforte** [comment]
hmm. so there is no way to repair a file replica, in spite of the documentation which I found earlier ?

**Comment by belforte** [comment]
No, that footnote is sitll [there](http://rucio.cern.ch/documentation/started/concepts/file_dataset_container#footnotes) also with new doc. Do you understand what it means then ? Can we ask Martin  ?

**Comment by dynamic-entropy** [comment]
Hmm, I have asked Martin and others for clarification.
 According to me, A file that is not binary identical would have to be registered as a separate file. And thus attached separately. 
 I will get back on this.

**Comment by belforte** [comment]
thanks you Rahul. This is extremely useful, even if I start to worry that we have been sitting for months on a plan which may be not possible to implement !
.

**Comment by dynamic-entropy** [comment]
Hi @belforte 
Quoting Martin's reply;

> I think this footnote was mostly meant for manual interventions. E g there have been cases where a file was fully lost, and it was removed from a closed dataset. Should a copy of this file still show up (let's say on a usb stick) an operator could still attach it to the dataset, although it is closed.
Probably this footnote should be removed, it really describes a rather obscure corner case. 

I still do not understand the `Binary identical` part. But in any case, I am sure that DID metadata cannot be changed. Not even manually by an operator.

**Comment by belforte** [comment]
thanks a lot @dynamic-entropy !!
Summin up:
* the *obscure corner case* seems a polite way to say "yeah, that could never work"
* a large number of "us" were utterly confused on replica properties vs. DID properties and ended up making plainly wrong assumptions
* I, as the one who introduced the idea that *the footnote* could save the day, was foolish not to have asked Martin directly

I believe that none of the fallback strategies for CRAB stageout discussed so far will be really appealing to users. We need some brainstorming and original thinking, possibly along the line which you suggested of "register the intial replica from grid job stageout as a permanent file" which you indicated.

I will be interested in any idea from anybody on this topic.

**Comment by dciangot** [comment]
I struggling to understand the point of the bad replica, I think the point where we left was about deleting replicas of a did. Did anyone can confirm that [this](https://github.com/dmwm/CMSRucio/issues/297#issuecomment-1274166707) still working ?

Meaning, delete replicas of a did and register the same did with a different replica/checksum?

Sorry to chime in late here, but chep was ongoing and quite packed.

**Comment by dciangot** [comment]
Well, at this point I start wondering if it is better to keep the fts transfer mechanism till the very end, and register the final replica directly on the destination RSE.

**Comment by belforte** [comment]
We need to be careful and write down all details.

The current flow is still fine. We are only discussing how to deal with exceptions which we do not even know how frequently will happen. At the end of the day, there's always the possibility of a new task which can recover whatever data is missing, and will create new LFN's. We may make it possible to run recovery based on info on Rucio, rather than DBS, @mapellidario 

In that scenario we would forbid resubmissions of jobs which reached the ASO step. The only thing we need is to declare a bad replica **in order to stop Rucio from trying to move a file which nobody will use**. That'd be quite simple to implement.

Or, if in the end "Rucio did not copy the file home" is very rare, oh well, maybe we introduce back `myout_10_1.root` a' la CRAB2. If we can make sure that `myout_10.root` is not on disk, nor in Rucio memory.. no confusion. The important thing is to guarantee a correct content for `/store/user/rucio/belforte/....../mydir`.

Generally speaking if we enable users to upload data in Rucio we should also give them a way to deal with lost or corrupted files by themselves.

Starting point is of course to know how things exactly work inside Rucio. We overlooked this step, let't take the time to make sure that we do not.

**Comment by ericvaandering** [comment]
This issue has migrated well beyond its initial purpose (and its initial purpose is now, irrelevant I would say). What should we do with this issue?

**Comment by belforte** [comment]
From my POV this can be closed. AFAIU the policy plugins offer a good way of taking car of needs in this area, as they surface.
Sorry for having hijacked this a bit into a (very useful by the way) digression into "Q: can DID's metadata be changes. A: No way"
